### PR TITLE
CMake Overhaul

### PR DESCRIPTION
**Pull request by fsimonis**
The PR features the overhaul of the CMake build configuration of precice.

I kept the _Interface_ stable, so you can configure using the old way:
```
$ cmake -DMPI=ON -DPETSC=ON -DPYTHON=ON \
        -DCMAKE_BUILD_TYPE=Debug \
        -DBUILD_SHARED_LIBS=ON \
         $PRECICE_ROOT
```
After that, you can:
1) Build `make`
2) Run tests `make test` or `ctest -V`
3) Generate packages `make package` or `cpack`
4) Install precice `[sudo] make install`

For a scoped install:
```
$ cmake -DMPI=ON -DPETSC=ON -DPYTHON=ON \
        -DCMAKE_BUILD_TYPE=Debug \
        -DBUILD_SHARED_LIBS=ON \
        -DCMAKE_INSTALL_PREFIX=/path/to/install/dir \
         $PRECICE_ROOT
$ make install -j $(nproc)
```

## Changes
  - New Options:
    - `PRECICE_Packages` (default ON) to configure CPack
    - `PRECICE_InstallTest` (default ON) to configure installation of tests.  
      This includes the binary `testprecice` and necessary files.
      Use `PREFIX/share/precice` as `PRECICE_ROOT`.
  - Move CMake files from `tools/cmake-modules` to `cmake/` (general scripts) and `cmake/modules` (find modules).
  - Migration from file-globbing to explicit source/interface/test-file lists.  
    Simply run the helper `tools/updateSourceFiles.py` from project-root to update all necessary files.
    __Rational:__ CMake evaluates file glob only during configuration time, adding a file to the source tree will thus not be picked up by the build system. [More Information](https://www.youtube.com/watch?v=bsXLMQ6WgIk&t=16m).
    
  - `install` target installs:
     - the library `PREFIX/lib`.
     - the binaries `PREFIX/bin` and their manfiles `PREFIX/share/man/man1`.
     - the cmake configuration files `PREFIX/lib/cmake/precice`.
     - the necessary files to run testprecice `PREFIX/share/precice`. Use this as `PRECICE_ROOT` on installed system.
  - CTest definition of tests running in isolated working directories:
    - `precice.Base` for the base test suite
    - `precice.MPI2` run on 2 MPI ranks
    - `precice.MPI4` run on 4 MPI ranks
  - CPack configuration of target `package` to generate binary debian, tar and zip packages.
  - Add `CMakeLists.txt` to `tools/solverdummy/cpp`. It is an example of how to link to precice with CMake.
  - Extend the displayed information when configuring.

## Solverdummy Cpp
This is now a tiny independent CMake project which aims to:
1. test the generated cmake configuration file for precice
2. test the installation of cmake _aka smoke test_
3. shows how to link to precice from a cmake project

If precice was installed somewhere unusual, use the following option in configuration:
```
$ cmake -DCMAKE_PREFIX_PATH="${CMAKE_PREFIX_PATH};/path/to/install/dir" $PRECICE_ROOT/tools/solverdummies/cpp
```

### CODE REVIEW

**Comment by floli** [comment]
> Migration from file-globbing to explicit source/interface/test-file lists.
> Simply run the helper tools/updateSourceFiles.py from project-root to update all necessary files.
> Rational: CMake evaluates file glob only during configuration time, adding a file to the source tree will thus not be picked up by the build system. More Information.

I have heard this rationale multiple times and I am ok with it. But it seems, now adding a file requires to rerun a script, then rerun cmake and make. Before it was just running cmake and make, not?

**Comment by Summerdave** [comment]
@floli
This is from a developer's perspective. Imagine you add a file between commits and don't alter `CMakeLists.txt`. A user will check out the commit and since `CMakeLists.txt` was not changed, `cmake` will not be re-run. _This is good, as it is not necessary_. In link stage the build will fail because the added file is not even compiled. So as a developer you run `updateSourceFiles.py`, `cmake` and `make` but only when you add a file. Else everyone just runs `make`.

File: tools/packaging/README.txt [hunk 1]
```diff
@@ -0,0 +1,10 @@
+preCICE stands for Precise Code Interaction Coupling Environment.
+
+Its main component is a library that can be used by simulation programs to be coupled together in a partitioned way, enabling multi-physics simulations, such as fluid-structure interaction.
+
+If you are new to preCICE, please have a look at our wiki https://github.com/precice/precice/wiki and at precice.org https://www.precice.org.
+
+preCICE is an academic project, developed at the Technical University of Munich https://www5.in.tum.de/ and at the University of Stuttgart https://www.ipvs.uni-stuttgart.de/.
```
Comment (uekerman) [inline tools/packaging/README.txt]:
```suggestion
preCICE is an academic project, developed at the Technical University of Munich https://www5.in.tum.de/, the University of Stuttgart https://www.ipvs.uni-stuttgart.de/, and the EIndhoven University of Technology https://www.tue.nl/en/research/research-groups/energy-technology/.
```

File: tools/packaging/debian/copyright [hunk 2]
```diff
@@ -0,0 +1,9 @@
+Format: http://www.debian.org/doc/packaging-manuals/copyright-format/1.0/
+Upstream-Name: preCICE
+Upstream-Contact: preCICE mailing-list <precice@mailman.informatik.uni-stuttgart.de>
+Copyright: (c) 2018 the preCICE developers <precice@mailman.informatik.uni-stuttgart.de>
```
Comment (uekerman) [inline tools/packaging/debian/copyright]:
```suggestion
Copyright: (c) 2019 the preCICE developers <precice@mailman.informatik.uni-stuttgart.de>
```

File: tools/packaging/debian/copyright [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+Source: https://github.com/precice/precice
+
+Files: *
+Copyright: (c) 2018 the preCICE developers <precice@mailman.informatik.uni-stuttgart.de>
```
Comment (uekerman) [inline tools/packaging/debian/copyright]:
```suggestion
Copyright: (c) 2019 the preCICE developers <precice@mailman.informatik.uni-stuttgart.de>
```

**Comment by fsimonis** [comment]
@floli @Summerdave After running the python script you can actually run make straight away. It will notice that the cmake sources have changed and automatically rerun the cmake configure step for you.

So the workflow is:
1) add cpp files
2) call `tools/updateSources.py`
3) `make`

File: tools/packaging/debian/copyright [hunk 4]
```diff
@@ -0,0 +1,9 @@
+Format: http://www.debian.org/doc/packaging-manuals/copyright-format/1.0/
+Upstream-Name: preCICE
+Upstream-Contact: preCICE mailing-list <precice@mailman.informatik.uni-stuttgart.de>
+Copyright: (c) 2019 the preCICE developers <precice@mailman.informatik.uni-stuttgart.de>
```
Comment (floli) [inline tools/packaging/debian/copyright]: My previous comment was shown as outdated, which it definitively not is. I think the mail address should be `info@precice.org`

File: tools/packaging/debian/copyright [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+Source: https://github.com/precice/precice
+
+Files: *
+Copyright: (c) 2019 the preCICE developers <precice@mailman.informatik.uni-stuttgart.de>
```
Comment (floli) [inline tools/packaging/debian/copyright]: Same as above, mail address `info@precice.org`

**Comment by floli** [comment]
Ok, now I see how the workflow goes.

**Pull request merged**